### ISSUE TITLE
Syntax error with a TS interface using syntax-typescript

### ISSUE THREAD

**Issue by mquandalle** [babel/babel#8099]
## Bug Report

**Current Behavior**
Syntax error with some TS interfaces and `@babel/plugin-syntax-typescript`

**Input Code**
```ts
interface myInterface  {
  method: ({test}: {test:string}) => string;
}
```

```
ERROR
Unexpected token, expected ")" (2:17)

 |   method: ({test}: {test:string}) => string;
 |                  ^
```

**Expected behavior/code**
The above code snippet should compile. The whole `interface` block should be removed by `@babel/plugin-syntax-typescript`.

**Environment**
- Babel version(s): v7.0.0-beta.42
- Node/npm version: Node 8/npm 5
- OS: Ubuntu 18.04

**Comment by mquandalle** [comment]
For information, this issue is still present in `v7.0.0-beta.51`

**Comment by alangpierce** [comment]
I just ran into this same issue in the Sucrase project (which has a fork of Babel's parser): https://github.com/alangpierce/sucrase/issues/277

It looks like the underlying issue here is destructuring within a parameter in a function type. The `tsIsStartOfFunctionType` method in babel-parser has a bunch of rules to determine when to take the function parsing code path, but returns `false` here when it should return `true`. Changing `({test}: {test:string}) => string` to `(args: {test:string}) => string` in the original code snippet fixes the issue.

[Repl link](http://babeljs.io/repl#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=MYewdgzgLgBAZgLhgCgN4EMC-SNLAVwFsAjAUwCdMBKGAXgD4YA3EASwBM6UaGZVMA3EA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&sourceType=module&lineWrap=false&presets=es2015%2Ctypescript&prettier=false&targets=&version=7.0.0-beta.42&envVersion=)

**Comment by noahsug** [comment]
Still present in `@babel/parser": "^7.0.0-beta.56`

**Comment by mmtjhlee** [comment]
It's still a thing. See [Minimal repro case in babel REPL](https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=PTAECUFMBMFcDtoEN4BdTUgZ1QJ1gMaqy4CW8A5qLpAA64D2oBSWkAUOapLgGZIFIoAGIMmAb1AAPAFyh4sALYAjHgG5QAX3ac0PfoNAAhJLlCT2oUFgaLIqABbkKcgBSSpWuaIYBKUAC8AHygAG4MpNBqlqAg1rb2TpRuvGLeYv7BYRFRsWAA6gDy4ADSAMrs2uyoAJ60Qia4WIHm8XaOzm4eXiIZgSHhkRqa0XG19camzQGtNu1JLqCuqQzpfv3ZQ1oacUWlFWN1QmWKji2Sc4mdSyty4rLySqq4mpkDOcN5oHvlQA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=typescript%2Cenv&prettier=false&targets=Node-8.9&version=7.0.0-beta.40&envVersion=7.0.0-rc.4) for causes and alternatives.

**Comment by sleagon** [comment]
Still present in released version @babel/parser": "^7.0.0"

**Comment by tommy-dev** [comment]
I also found this bug while using gatsbyjs version2, which presumably uses some babel 7 parser.

type Props = {
    onSelect?: ({ value: any, on: boolean }) => void                  <=error
}
